.github/workflows/commit-tests.yml

```diff
@@ -0,0 +1,137 @@
+name: Single Test
+on:
+  push:
+
+permissions: read-all
+
+env:
+  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+jobs:
+  parse-commit:
+    runs-on: ubuntu-latest
+    # if the commit message does not contain {craft} or {bash} tags, then skip the craft-test and bash-test jobs
+    if: github.event_name == 'push' && (contains(github.event.head_commit.message, '{craft}') || contains(github.event.head_commit.message, '{bash}'))
+    outputs:
+      skip_craft: ${{ steps.parse_commit.outputs.skip_craft }}
+      craft: ${{ steps.parse_commit.outputs.craft }}
+      skip_bash: ${{ steps.parse_commit.outputs.skip_bash }}
+      bash: ${{ steps.parse_commit.outputs.bash }}
+      build_os: ${{ steps.parse_commit.outputs.build_os }}
+      spc_prefix: ${{ steps.parse_commit.outputs.spc_prefix }}
+    steps:
+      - name: "Checkout"
+        uses: actions/checkout@v4
+
+      - name: "Parse commit message"
+        id: parse_commit
+        run: |
+          # parse the commit message, see if it has {craft} and {/craft} tags
+          COMMIT_MESSAGE=$(git log -1 --pretty=%B)
+          # judge it, it it's not exist, then skip this test
+          if [[ "$COMMIT_MESSAGE" != *"{craft}"* ]] || [[ "$COMMIT_MESSAGE" != *"{/craft}"* ]]; then
+            echo "No {craft} tags found in commit message. Skipping test."
+            echo "skip_craft=yes" >> $GITHUB_OUTPUT
+            exit 0
+          else
+            echo -e "\e[32mCraft tags found in commit message.\e[0m"
+            # get the craft content
+            CRAFT_CONTENT=$(echo "$COMMIT_MESSAGE" | sed -nz 's/.*{craft}\(.*\){\/craft}.*/\1/p')
+            echo "Craft content: $CRAFT_CONTENT"
+            # set the output variable
+            echo "craft<<EOF" >> $GITHUB_OUTPUT
+            echo "$CRAFT_CONTENT" >> $GITHUB_OUTPUT
+            echo "EOF" >> $GITHUB_OUTPUT
+          fi
+
+          # parse the bash test script from the commit message
+          if [[ "$COMMIT_MESSAGE" != *"{bash}"* ]] || [[ "$COMMIT_MESSAGE" != *"{/bash}"* ]]; then
+            echo "No {bash} tags found in commit message. Skipping bash test."
+            echo "skip_bash=yes" >> $GITHUB_OUTPUT
+          else
+            echo -e "\e[32mBash tags found in commit message.\e[0m"
+            # get the bash content
+            BASH_CONTENT=$(echo "$COMMIT_MESSAGE" | sed -nz 's/.*{bash}\(.*\){\/bash}.*/\1/p')
+            echo "Bash content: $BASH_CONTENT"
+            # set the output variable
+            echo "bash<<EOF" >> $GITHUB_OUTPUT
+            echo "$BASH_CONTENT" >> $GITHUB_OUTPUT
+            echo "EOF" >> $GITHUB_OUTPUT
+          fi
+
+          # parse spc_prefix from commit message, e.g. [spc_prefix:bin/spc-gnu-docker], default: bin/spc
+          if [[ "$COMMIT_MESSAGE" =~ \[spc_prefix:([^\]]+)\] ]]; then
```

Comment:
I think this is too difficult to actually use. To test with everything we'd need to define "craft bash spc_prefix=spc-gnu-docker spc_prefix`?

Reply:
To be honest, I also feel that it is a bit complicated, but I can't find a balanced solution.

I also thought about combining ext-matrix-test with test-extensions.php. When the message contains "[matrix:8.4:macos]", test all extensions on the 8.4 build of mac, but still need to define a set of simple and easy-to-use tag rules.

Directly implanting craft is the easiest way to use the ready-made solution without defining the syntax. But if I don't consider the syntax at all, I can't think of a more optimized solution to dynamically trigger the build.

Reply:
Hahaha, surely that would go well!

> "[matrix:8.4:macos]"

Again, to complicated for my liking. May be fine for us, but it's hard to eventually tell people how to properly test their code.

We don't need to be that selective about testing, I think. Just trigger *all* tests when the commit message contains "fix" or "test". The whole matrix with all extensions/libraries currently defined in test-extensions.